internal/transport/grpchttp2/framer.go

```diff
@@ -0,0 +1,252 @@
+	// StreamID is the ID for the stream which this frame is for. If the
+	// frame is connection specific instead of stream specific, the
+	// streamID is 0.
+	StreamID uint32
+}
+
+// Frame represents an HTTP/2 Frame.
+type Frame interface {
+	Header() *FrameHeader
+	// Free frees the underlying buffer if present so it can be reused by the
+	// framer.
+	Free()
+}
+
+type DataFrame struct {
+	hdr  *FrameHeader
+	free func()
+	Data []byte
+}
+
+func (f *DataFrame) Header() *FrameHeader {
+	return f.hdr
+}
+
+func (f *DataFrame) Free() {
+	if f.free != nil {
+		f.free()
+	}
+}
+
+type HeadersFrame struct {
+	hdr      *FrameHeader
+	free     func()
+	HdrBlock []byte
+}
+
+func (f *HeadersFrame) Header() *FrameHeader {
+	return f.hdr
+}
+
+func (f *HeadersFrame) Free() {
+	if f.free != nil {
+		f.free()
+	}
+}
+
+type RSTStreamFrame struct {
+	hdr  *FrameHeader
+	Code ErrCode
+}
+
+func (f *RSTStreamFrame) Header() *FrameHeader {
+	return f.hdr
+}
+
+func (f *RSTStreamFrame) Free() {}
+
+type SettingsFrame struct {
+	hdr      *FrameHeader
+	Settings []Setting
+}
+
+func (f *SettingsFrame) Header() *FrameHeader {
+	return f.hdr
+}
+
+func (f *SettingsFrame) Free() {}
+
+type PingFrame struct {
+	hdr  *FrameHeader
+	free func()
+	Data []byte
+}
+
+func (f *PingFrame) Header() *FrameHeader {
+	return f.hdr
+}
+
+func (f *PingFrame) Free() {
+	if f.free != nil {
+		f.free()
+	}
+}
+
+type GoAwayFrame struct {
+	hdr          *FrameHeader
+	free         func()
+	LastStreamID uint32
+	Code         ErrCode
+	DebugData    []byte
+}
+
+func (f *GoAwayFrame) Header() *FrameHeader {
+	return f.hdr
+}
+
+func (f *GoAwayFrame) Free() {
+	if f.free != nil {
+		f.free()
+	}
+}
+
+type WindowUpdateFrame struct {
+	hdr *FrameHeader
+	Inc uint32
+}
+
+func (f *WindowUpdateFrame) Header() *FrameHeader {
+	return f.hdr
+}
+
+type ContinuationFrame struct {
+	hdr      *FrameHeader
+	free     func()
+	HdrBlock []byte
+}
+
+func (f *ContinuationFrame) Header() *FrameHeader {
+	return f.hdr
+}
+
+func (f *ContinuationFrame) Free() {
+	if f.free != nil {
+		f.free()
+	}
+}
+
+// MetaHeadersFrame is not a Frame Type that appears on the HTTP/2 Spec. It is
+// a representation of the merging and decoding of all the Headers and
+// Continuation frames on a Stream.
+type MetaHeadersFrame struct {
+	hdr    *FrameHeader
+	Fields []hpack.HeaderField
+}
+
+func (f *MetaHeadersFrame) Header() *FrameHeader {
+	return f.hdr
+}
+
+func (f *MetaHeadersFrame) Free() {}
+
+// Framer encapsulates the functionality to read and write HTTP/2 frames.
+type Framer interface {
+	// SetMetaDecoder will set a decoder for the framer. When the decoder is
+	// set, ReadFrame will parse the header values, merging all Headers and
+	// Continuation frames.
+	SetMetaDecoder(d *hpack.Decoder)
+	// ReadFrame returns an HTTP/2 Frame. It is the caller's responsibility to
+	// free the frame once it is done using it.
+	ReadFrame() (Frame, error)
+	WriteData(streamID uint32, endStream bool, data ...[]byte) error
```

Comment:
For the time being could we add docstring to each of these methods?  It's easier to see these docstrings while going through the code in the interim because calls from gRPC would always lead us to this interface definition until we only have one concrete type.